Variable declaration

333fred opened the pull request:
Adds support for IVariableDeclarationGroup, IVariableDeclaration, and IVariableDeclarator to control flow graph generation.

mavasani commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs:
```diff
@@ -1190,15 +1190,14 @@ void M()
-                      ILiteralOperation (OperationKind.Literal, Type: System.Boolean, Constant: False) (Syntax: 'false')
-            Initializer: 
-              null
+        IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'condition = false')
```
I recall us talking to rewrite variable initialization into simple assignment, but I am not sure why are we removing the IVariableDeclaration operations itself? I am not sure if this aids any further in control flow analysis. In fact, if some client wants to do some variable analysis and report a diagnostic on the variable declaration, removing the `IVariableDeclarationOperation` nodes makes it difficult for them.

333fred replied:
The IVariableDeclaration and the initializer aren't guaranteed to be in the same place, so clients will have to handle this anyway. If the client needs to report a diagnostic on the symbol's original definition, they'd have to go through `DeclaringSyntaxReferences` anyway.

mavasani replied:
Yes, but I am not sure what is the value add or necessity of removing `IVariableDeclaration` operation for flow analysis?

333fred replied:
There's no benefit to leaving it in, and it adds unnecessary nodes to the graph. We've already said that we're going to have variable lifetime information in a separate data structure, and all the `IVariableDeclaration` does (or really `IVariableDeclarator`, which is what I think you mean here) is provide that information.

333fred replied:
@jinujoseph, we'll bring this up in the meeting today.

333fred replied:
Per design discussion today, we'll fully remove the IVariableDeclarat(ion(Group))(or) from the resulting graph.

mavasani commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs:
```diff
@@ -1190,15 +1190,14 @@ void M()
+              Left: 
+                ILocalReferenceOperation: condition (IsDeclaration: True) (OperationKind.LocalReference, Type: System.Boolean, IsImplicit) (Syntax: 'condition = false')
+              Right: 
+                IVariableInitializerOperation (OperationKind.VariableInitializer, Type: null) (Syntax: '= false')
```
Why are we retaining the `IVariableInitializerOperation`, which is normally expected to be parented by a variable declaration? I thought we were going to do the reverse lowering - remove `IVariableInitializerOperation` nodes and add subsequent simple assignments operations for initialization.

mavasani replied:
Note that not having to deal with different kind of initializers (simple variable/param/member initializer, array initializer, object initializer, collection initializer, member initializer, etc) does simplify flow analysis significantly, and lowering them into simple assignments followed by a declaration is definitely a value add.

333fred replied:
My plan this PR was to only handle `IVariableDeclarationGroup`, `IVariableDeclaration`, and `IVariableDeclarator`. If reviewing it is easier, I can also add in `IVariableInitializerOperation`, but I figured that would be handled in the same PR as the other initializer operations.

mavasani replied:
If lowering `IVariableInitializerOperation` is coming in next PR, that is fine. However, I am still not convinced why get rid of the declaration operations. What is the motivating factor behind this change?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,65 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+                SyntaxNode assignmentSyntax = null;
+                if (declarator.Initializer != null)
+                {
+                    initializer = Visit(declarator.Initializer);
```
>declarator.Initializer [](start = 40, length = 22)

We should skip through IVariableInitializerOperation to the Value here

333fred replied:
I'm planning on doing that skip in a separate PR for all the initializer types. I don't think the skip though should be here, the visitor for IVariableInitializer should just remove itself from the tree and return visit of it's child.

---
In reply to: [166775123](https://github.com/dotnet/roslyn/pull/24692#discussion_r166775123) [](ancestors = 166775123)

AlekseyTs replied:
It should be skipped here and in context of this PR. IVariableInitializerOperation doesn't represent a value and doesn't belong on the right hand side of the assignment. The remaining initializer types can wait as we haven't even decided what shape should the graph have for them and do not have an API for requesting it yet.

---
In reply to: [166777002](https://github.com/dotnet/roslyn/pull/24692#discussion_r166777002) [](ancestors = 166777002,166775123)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,65 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+
+        public override IOperation VisitVariableDeclaration(IVariableDeclarationOperation operation, int? captureIdForResult)
+        {
+            foreach (var declarator in operation.Declarators)
```
>var [](start = 21, length = 3)

Please no ```var```s. The type must be spelled out unless it is spelled out explicitly in the initializer, which is not present here.

333fred replied:
Fixed.

---
In reply to: [166775498](https://github.com/dotnet/roslyn/pull/24692#discussion_r166775498) [](ancestors = 166775498)

AlekseyTs commented:
>             return new VariableInitializer(Visit(operation.Value), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);

We probably should throw unreachable here.

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1265 in 43561df. [](commit_id = 43561df0b021508ede099c7dd319f22f8d27ee0e, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,65 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+
+                if (operation.Initializer != null)
+                {
+                    IOperation operationInitializer = Visit(operation.Initializer);
```
>operation.Initializer [](start = 60, length = 21)

Same here, IVariableInitializerOperation should be dropped completely.

333fred commented:
>             return new VariableInitializer(Visit(operation.Value), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);

I'd rather have this and all the other initializer types just return `Visit(Operation.Value)`, less assumptions that need to be made.

---
In reply to: [363931562](https://github.com/dotnet/roslyn/pull/24692#issuecomment-363931562) [](ancestors = 363931562)

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1265 in 43561df. [](commit_id = 43561df0b021508ede099c7dd319f22f8d27ee0e, deletion_comment = False)

AlekseyTs commented:
>             return new VariableInitializer(Visit(operation.Value), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);

> I'd rather have this and all the other initializer types just return Visit(Operation.Value), less assumptions that need to be made.

I don't think this is the right approach. This node doesn't represent a value and we should not be changing this during regular Visit.

---
In reply to: [363933264](https://github.com/dotnet/roslyn/pull/24692#issuecomment-363933264) [](ancestors = 363933264,363931562)

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1265 in 43561df. [](commit_id = 43561df0b021508ede099c7dd319f22f8d27ee0e, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,65 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+                    // We can't use the IdentifierToken as the syntax for the local reference, so we use the entire declarator as the node
+                    var localRef = new LocalReferenceExpression(localSymbol, isDeclaration: true, semanticModel: null, declarator.Syntax, localSymbol.Type, constantValue: default, isImplicit: true);
+
+                    var statement = new ExpressionStatement(new SimpleAssignmentExpression(localRef, isRef: localSymbol.IsRef, initializer, semanticModel: null, assignmentSyntax, localRef.Type, constantValue: default, isImplicit: true),
```
>new ExpressionStatement [](start = 36, length = 23)

I think we can skip creation of an ExpressionStatement here. It is fine to use simple assignment as a statement.

333fred replied:
I considered this, but my thought was that for analyzer authors it would be helpful if:
```C#
int a = 1;
int b;
b = 1;
```
ended up with the same shape in the flow graph, namely an IExpressionStatementOperation wrapping an ISimpleAssignmentOperation.

---
In reply to: [166782118](https://github.com/dotnet/roslyn/pull/24692#discussion_r166782118) [](ancestors = 166782118)

AlekseyTs replied:
With IOperation from syntax our hand was forced by syntax. There is no constraint like that for synthetic nodes, and I think we should keep amount synthesized nodes to a necessary minimum

---
In reply to: [166783230](https://github.com/dotnet/roslyn/pull/24692#discussion_r166783230) [](ancestors = 166783230,166782118)

333fred replied:
I'm going to leave this as is for now. During design discussion next week, we'll talk about removing all ExpressionStatements from the Control Flow graph, as they don't really serve a purpose.

---
In reply to: [166784473](https://github.com/dotnet/roslyn/pull/24692#discussion_r166784473) [](ancestors = 166784473,166783230,166782118)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,65 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+            return null;
+        }
+
+        public override IOperation VisitVariableDeclaration(IVariableDeclarationOperation operation, int? captureIdForResult)
```
>public override IOperation VisitVariableDeclaration(IVariableDeclarationOperation operation, int? captureIdForResult) [](start = 8, length = 117)

```private void VisitVariableDeclaration(IVariableDeclarationOperation operation)```? And throw from ```public override IOperation VisitVariableDeclaration(IVariableDeclarationOperation operation, int? captureIdForResult)```

333fred replied:
Will do.

---
In reply to: [166783214](https://github.com/dotnet/roslyn/pull/24692#discussion_r166783214) [](ancestors = 166783214)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,65 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+                VisitVariableDeclaration(declaration, captureIdForResult);
+            }
+
+            return null;
```
>return null; [](start = 12, length = 12)

Is IVariableDeclarationGroupOperation always used as a statement? Consider asserting that it is equal to the ```_currentStatement```.

333fred replied:
It's not used as the statement for `for` loops.

---
In reply to: [166783785](https://github.com/dotnet/roslyn/pull/24692#discussion_r166783785) [](ancestors = 166783785)

AlekseyTs replied:
> It's not used as the statement for for loops.

It well be. We are not handling loops properly at the moment. Please add an assert.

---
In reply to: [166784997](https://github.com/dotnet/roslyn/pull/24692#discussion_r166784997) [](ancestors = 166784997,166783785)

AlekseyTs replied:
To clarify, we cannot return null, unless we are visiting this node in a statement mode.

---
In reply to: [167016477](https://github.com/dotnet/roslyn/pull/24692#discussion_r167016477) [](ancestors = 167016477,166784997,166783785)

333fred replied:
Added.

---
In reply to: [167018923](https://github.com/dotnet/roslyn/pull/24692#discussion_r167018923) [](ancestors = 167018923,167016477,166784997,166783785)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,65 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
             return null;
         }
 
+        public override IOperation VisitVariableDeclarationGroup(IVariableDeclarationGroupOperation operation, int? captureIdForResult)
```
>VisitVariableDeclarationGroup [](start = 35, length = 29)

Please make necessary changes to ControlFlowGraphVerifier.CanBeInControlFlowGraph method to make sure that nodes that you've handled are in the right set. It is fine that there still can be scenarios that we do not handle, but as long as all current tests are passing, it would be good to accurately reflect final expectations in that function.

333fred replied:
Updated.

---
In reply to: [166806899](https://github.com/dotnet/roslyn/pull/24692#discussion_r166806899) [](ancestors = 166806899)

333fred commented:
@mavasani @AlekseyTs updated to address feedback.

mavasani commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,76 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+
+                if (operation.Initializer != null)
+                {
+                    IOperation operationInitializer = Visit(operation.Initializer.Value, captureIdForResult);
```
Will we visit the `operation.Initializer` multiple times here for a declaration with multiple declarators sharing the same initializer? I feel we should visit them only once, and re-use the visited result for all assignments (probably move `operationInitializer` declaration outside the loop, and execute the visit only if `operationInitializer == null`?)

AlekseyTs replied:
We cannot reuse the node as it should have a different parent in each assignment

---
In reply to: [166980240](https://github.com/dotnet/roslyn/pull/24692#discussion_r166980240) [](ancestors = 166980240)

AlekseyTs replied:
Also, Value can contain control flow, which will add statements and basic blocks, simply reusing the returned node will not add these "side-effects".

---
In reply to: [167017796](https://github.com/dotnet/roslyn/pull/24692#discussion_r167017796) [](ancestors = 167017796,166980240)

AlekseyTs replied:
@333fred Please add a PROTOTYPE comment that we should add a test with a control flow in a shared initializer. We can add it right now, but it I think it won't produce expected result because we don't handle control flow in object creation yet. 

---
In reply to: [167020856](https://github.com/dotnet/roslyn/pull/24692#discussion_r167020856) [](ancestors = 167020856,167017796,166980240)

333fred replied:
Added.

---
In reply to: [167023541](https://github.com/dotnet/roslyn/pull/24692#discussion_r167023541) [](ancestors = 167023541,167020856,167017796,166980240)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,76 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+                SyntaxNode assignmentSyntax = null;
+                if (declarator.Initializer != null)
+                {
+                    initializer = Visit(declarator.Initializer.Value, captureIdForResult);
```
>captureIdForResult [](start = 70, length = 18)

We should not be propagating ```captureIdForResult``` here. We are not producing a value here, the parameter should be removed from the calls and the enclosing function.

333fred replied:
Removed.

---
In reply to: [167019822](https://github.com/dotnet/roslyn/pull/24692#discussion_r167019822) [](ancestors = 167019822)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,76 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+                {
+                    // We can't use the IdentifierToken as the syntax for the local reference, so we use the entire declarator as the node
+                    var localRef = new LocalReferenceExpression(localSymbol, isDeclaration: true, semanticModel: null, declarator.Syntax, localSymbol.Type, constantValue: default, isImplicit: true);
+                    // prototype(dataflow): We'd like to remove ExpressionStatements from the CFG altogether, as they're useless when all you need to do is look to see if the parent is null
```
>prototype [](start = 23, length = 9)

"prototype" should be in all caps

333fred replied:
Fixed all instances of this.

---
In reply to: [167021108](https://github.com/dotnet/roslyn/pull/24692#discussion_r167021108) [](ancestors = 167021108)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,76 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
+        public override IOperation VisitVariableDeclarator(IVariableDeclaratorOperation operation, int? captureIdForResult)
+        {
+            // All variable declarators should be handled by VisitVariableDeclaration.
+            // prototype(dataflow): try/catch can get directly here, as can foreach control variables
```
>// prototype(dataflow): try/catch can get directly here, as can foreach control variables [](start = 12, length = 89)

Please remove this comment. try/catch won't get here once we handle it properly. No adjustments will  be necessary here.

333fred replied:
Ok, removed.

---
In reply to: [167021700](https://github.com/dotnet/roslyn/pull/24692#discussion_r167021700) [](ancestors = 167021700)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -997,6 +1052,7 @@ public override IOperation VisitForToLoop(IForToLoopOperation operation, int? ca
 
         public override IOperation VisitForEachLoop(IForEachLoopOperation operation, int? captureIdForResult)
         {
+            // prototype(dataflow): note that the loop control variable can be an IVariableDeclarator directly, and this function is expected to handle it without calling Visit(IVariableDeclarator)
```
>prototype [](start = 15, length = 9)

All caps for "prototype" please.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs:
```diff
@@ -2288,5 +2288,580 @@ static void Main(string[] args)
+    void M(bool a, int b, int c)
+    /*<bind>*/{
+#pragma warning disable CS0219 // Variable is assigned but its value is never used
+        int d = a ? b : c;
```
>int d = a ? b : c [](start = 8, length = 17)

It would be good to have a test with two locals initialized in the same statement and control flow in the second initializer.

333fred replied:
Added for C# and VB.

---
In reply to: [167024883](https://github.com/dotnet/roslyn/pull/24692#discussion_r167024883) [](ancestors = 167024883)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs:
```diff
@@ -2288,5 +2288,565 @@ static void Main(string[] args)
+    Predecessors (1)
+        [0]
+    Statements (1)
+        IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'd = 1')
```
>IExpressionStatementOperation [](start = 8, length = 29)

Please add a PROTOTYPE comment, we need to discuss at design meeting whether we want to add assignments to constant locals at all. The fact that the symbol itself returns constant value could be sufficient.

333fred replied:
Added. FYI @jinujoseph for the design question.

---
In reply to: [167026093](https://github.com/dotnet/roslyn/pull/24692#discussion_r167026093) [](ancestors = 167026093)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs:
```diff
@@ -2288,5 +2288,565 @@ static void Main(string[] args)
+            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics);
+        }
+
+        // prototype(dataflow): test using/for/fixed constructs after support has been added.
```
>prototype [](start = 11, length = 9)

All caps?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb:
```diff
@@ -2765,6 +2765,637 @@ IVariableDeclarationOperation (2 declarators) (OperationKind.VariableDeclaration
+
+        IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'b As Integer = 1')
+          Expression: 
+            ISimpleAssignmentOperation (OperationKind.SimpleAssignment, Type: System.Int32, IsImplicit) (Syntax: 'b As Integer = 1')
```
>ISimpleAssignmentOperation [](start = 12, length = 26)

Please add a PROTOTYPE comment to confirm whether this node is good enough to indicate that this initialization of a static local happens only once, across all invocations of containing method

333fred replied:
Added. FYI @jinujoseph for the design question.

---
In reply to: [167030284](https://github.com/dotnet/roslyn/pull/24692#discussion_r167030284) [](ancestors = 167030284)

AlekseyTs commented:
Done with review pass (iteration 3).

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [364210264](https://github.com/dotnet/roslyn/pull/24692#issuecomment-364210264) [](ancestors = 364210264)

333fred commented:
Ping @dotnet/roslyn-compiler for a second review.

cston commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs:
```diff
@@ -2288,5 +2288,642 @@ static void Main(string[] args)
+            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics);
+        }
+
+        // PROTOTYPE(dataflow): test using/for/fixed constructs after support has been added.
```
Consider testing a variable assigned before it is declared.

333fred replied:
Added.

---
In reply to: [167055452](https://github.com/dotnet/roslyn/pull/24692#discussion_r167055452) [](ancestors = 167055452)

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -907,6 +907,79 @@ public override IOperation VisitWhileLoop(IWhileLoopOperation operation, int? ca
 
+        public override IOperation VisitVariableDeclarationGroup(IVariableDeclarationGroupOperation operation, int? captureIdForResult)
+        {
+            // Anything that has a declaration group (such as for loops) needs to handle them directly itself, this should only be encountered by the visitor for declaration statements.
```
Consider splitting the comment across multiple lines. Same for other long lines in this method.

333fred replied:
The "advantages" of the new monitors.

---
In reply to: [167056137](https://github.com/dotnet/roslyn/pull/24692#discussion_r167056137) [](ancestors = 167056137)
